crates/dojo-lang/src/interface.rs

```diff
@@ -0,0 +1,149 @@
+                    ("body".to_string(), RewriteNode::new_modified(body_nodes)),
+                ]),
+            ));
+
+            return PluginResult {
+                code: Some(PluginGeneratedFile {
+                    name: name.clone(),
+                    content: builder.code,
+                    aux_data: Some(DynGeneratedFileAuxData::new(DojoAuxData {
+                        models: vec![],
+                        systems: vec![SystemAuxData {
+                            name,
+                            dependencies: system.dependencies.values().cloned().collect(),
+                        }],
+                    })),
+                    code_mappings: builder.code_mappings,
+                }),
+                diagnostics: system.diagnostics,
+                remove_original_item: true,
+            };
+        }
+
+        PluginResult::default()
+    }
+
+    pub fn rewrite_function(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        fn_ast: ast::TraitItemFunction,
+    ) -> Vec<RewriteNode> {
+        let mut rewrite_nodes = vec![];
+
+        // Some notes:
+        // * as rewritten functions belong to a starknet::interface:
+        //     - there is no generic parameter,
+        //     - there is no nopanic and implicits clauses.
+        // * there is no function body in a trait.
+
+        let declaration = fn_ast.declaration(db);
+        let fn_name = declaration.name(db).text(db).to_string();
+        let signature = declaration.signature(db);
+
+        let attributes;
+        if fn_ast.attributes(db).elements(db).is_empty() {
+            attributes = "".to_string();
+        }
+        else {
+            attributes = fn_ast.attributes(db).elements(db)
+            .iter()
+            .map(|e| e.as_syntax_node().get_text(db))
+            .collect::<Vec<_>>()
+            .join("\n");
+        }
+
+        let ret_ty = if let OptionReturnTypeClause::ReturnTypeClause(ty) = signature.ret_ty(db) {
+            format!(" {}", ty.as_syntax_node().get_text(db).to_string())
+        } else {
+            "".to_string()
+        };
+
+        let mut params = signature.parameters(db).elements(db)
+            .iter()
+            .map(|e| e.as_syntax_node().get_text(db))
+            .collect::<Vec<_>>();
+
+        let need_to_add_self;
+        if params.is_empty() {
+            need_to_add_self = true;
+        }
+        else {
+            let first_param = signature.parameters(db).elements(db)[0].clone();
+            let first_param_name = first_param
+                .name(db).text(db)
+                .to_string();
+            let first_param_type = first_param
+                .type_clause(db).ty(db).as_syntax_node()
+                .get_text(db).trim().to_string();
+
+            need_to_add_self = !(
+                first_param_name.eq(&"self".to_string()) 
+                && first_param_type.eq(&"@TContractState".to_string())
+            );
```

Comment:
Wondering if here we should emit a warning (or error?) if the function is using `ref self`.
In dojo, systems are supposed to be 100% stateless. For this reason, we should consider warning/send error in this particular case.

Reply:
Yes, good idea !